src/rgen/Microsoft.Macios.Bindings.Analyzer/NativeObjectHandleAnalyzer.cs

```diff
@@ -0,0 +1,134 @@
+using Microsoft.CodeAnalysis.Diagnostics;
+using System.Collections.Immutable;
+using System.Linq;
+
+namespace Microsoft.Macios.Bindings.Analyzer;
+
+[DiagnosticAnalyzer (LanguageNames.CSharp)]
+public class NativeObjectHandleAnalyzer : DiagnosticAnalyzer {
+	internal static readonly DiagnosticDescriptor RBI0014 = new (
+		"RBI0014",
+		new LocalizableResourceString (nameof (Resources.RBI0014Title), Resources.ResourceManager, typeof (Resources)),
+		new LocalizableResourceString (nameof (Resources.RBI0014MessageFormat), Resources.ResourceManager,
+			typeof (Resources)),
+		"Usage",
+		DiagnosticSeverity.Error,
+		isEnabledByDefault: true,
+		description: new LocalizableResourceString (nameof (Resources.RBI0014Description), Resources.ResourceManager,
+			typeof (Resources))
+	);
+
+	public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create (RBI0014);
+
+	public override void Initialize (AnalysisContext context)
+	{
+		context.ConfigureGeneratedCodeAnalysis (GeneratedCodeAnalysisFlags.None);
+		context.EnableConcurrentExecution ();
+		context.RegisterSyntaxNodeAction (AnalyzeNode, SyntaxKind.SimpleMemberAccessExpression);
+	}
+
+	private bool IsINativeObject (ITypeSymbol typeSymbol)
+	{
+		return typeSymbol.AllInterfaces.Any (i => i.ToDisplayString () == "ObjCRuntime.INativeObject");
+	}
+
+	private bool IsSafeMethod (IMethodSymbol methodSymbol)
+	{
+		var name = methodSymbol.Name;
+		// White-list couple of methods that return one of the input parameters or a static
+		// object.
+		return name == "ThrowOnNull" || name == "DangerousAutorelease" || name == "GetConstant";
+	}
+
+	private void AnalyzeNode (SyntaxNodeAnalysisContext context)
+	{
+		if (context.Node is not MemberAccessExpressionSyntax memberAccess)
+			return;
+
+		if (memberAccess.Name.Identifier.Text != "Handle" &&
+			memberAccess.Name.Identifier.Text != "GetHandle" &&
+			memberAccess.Name.Identifier.Text != "GetNonNullHandle" &&
+			memberAccess.Name.Identifier.Text != "GetCheckedHandle")
+			return;
+
+		var symbol = context.SemanticModel.GetSymbolInfo (memberAccess.Expression).Symbol;
+		if (symbol is null)
+			return;
+
+		if ((symbol as IParameterSymbol)?.IsThis == true)
+			return;
+
+		if (symbol is IMethodSymbol methodSymbol) {
+			if (IsINativeObject (methodSymbol.ReturnType) &&
+				!IsSafeMethod (methodSymbol)) {
+				// Calling Handle directly on a value returned from method is wrong because
+				// the object is immediately collectible.                
+				var diagnostic = Diagnostic.Create (RBI0014, memberAccess.GetLocation (), symbol.Name);
+				context.ReportDiagnostic (diagnostic);
+			}
+			return;
+		}
+
+		var type = (symbol as ILocalSymbol)?.Type ?? (symbol as IParameterSymbol)?.Type;
+		if (type is null || !IsINativeObject (type))
+			return;
+
+		// Ignore variables that are wrapped in using block
+		var usingStatement = memberAccess.Parent?.FirstAncestorOrSelf<UsingStatementSyntax> ();
```

Comment:
We also need to check for LocalDeclarationStatementSyntax that contains a using, right? We will need to look for a LocalDeclarationStatementSyntax whose UsingKeyword is set.

Reply:
Do you mean to account for `using var x = y;` type of statements that get disposed at the end of the method? Ideally we should handle that too but realistically there were less than 5 places where the pattern was used so it was easier to just rewrite it to old-school `using` block.